### ISSUE TITLE
Inference with noisy source

### ISSUE THREAD

**Issue by Charlottecuc** [yl4579/StarGANv2-VC#21]
Hi. I tested the model with various kinds of wave files as source. I notice that at inference time, the model performs well with clean source files, but for those not so clean audio files (e.g. 24khz speech recorded by mobile phone, with background of air conditioning, or heavy breathing, which is quite common in real life application), the converted speech is sometime incomprehensible and usually with annoying noise.

I also tried denosing these noisy source files (e.g. using Audition, or other speech enhancement tools), but the converted speech became even worse. 

Besides, do you think this line `mel_tensor = (torch.log(1e-5 + mel_tensor) - self.mean) / self.std` to some extent enlarges the noise...?

Could you please give some ideas of making the model more robust with noisy data? Thank you very much.

**Comment by yl4579** [comment]
The normalization may have possibly amplified the noises slightly, yet the point of log mel spec is actually the opposite: it tries to emphasize the speech than noise. Probably the arbitrary mean and standard deviation may have some side effect, but if during training you make the model noise-robust, it should have no problems taking noisy input. As mentioned [here](https://github.com/yl4579/StarGANv2-VC/issues/6) earlier, if you corrupt your input with [Audiomentations](https://github.com/iver56/audiomentations), it should have no problems dealing with noisy input. Just make sure you separate your `x_real` and `x_input` and only make `x_input` noisy.

**Comment by Charlottecuc** [comment]
@yl4579 Hi. Thank you for your reply. 
Could you give any advice on the percentage of noisy input? Or, should all the `x_input` be corrupted? 
I notice that you mentioned "you can use the contrastive loss to make it noise-robust. That is, generate a noise degenerated copy of your audio and make the style encoder encode both of them into the same style vector". Should I only change the `x_input` of style encoder part? Thank you very much

**Comment by Charlottecuc** [comment]
e.g. 
`def compute_g_loss(nets, args, x_real, y_org, y_trg, z_trgs=None, x_refs=None, use_adv_cls=False, x_noise=None):
    args = Munch(args)
    
    assert (z_trgs is None) != (x_refs is None)
    if z_trgs is not None:
        z_trg, z_trg2 = z_trgs
    if x_refs is not None:
        x_ref, x_ref2 = x_refs
        
    # compute style vectors
    if z_trgs is not None:
        s_trg = nets.mapping_network(z_trg, y_trg)
    else:
        s_trg = nets.style_encoder(x_ref, y_trg)
    
    # compute ASR/F0 features (real)
    with torch.no_grad():
        F0_real, GAN_F0_real, cyc_F0_real = nets.f0_model(x_real)
        ASR_real = nets.asr_model.get_feature(x_real)
        F0_noise, GAN_F0_noise, cyc_F0_noise = nets.f0_model(x_noise)
    
    # adversarial loss
    x_fake_noise = nets.generator(x_noise, s_trg, masks=None, F0=GAN_F0_noise)
    x_fake_clean = nets.generator(x_real, s_trg, masks=None, F0=GAN_F0_real)
    out = nets.discriminator(x_fake_noise, y_trg) 
    loss_adv = adv_loss(out, 1)
    
    # compute ASR/F0 features (fake)
    F0_fake_noise, GAN_F0_fake_noise, _ = nets.f0_model(x_fake_noise)
    ASR_fake_noise = nets.asr_model.get_feature(x_fake_noise)
    
    F0_fake_clean, GAN_F0_fake_clean, _ = nets.f0_model(x_fake_clean)
    ASR_fake_clean = nets.asr_model.get_feature(x_fake_clean)
    
    # norm consistency loss
    x_fake_norm = log_norm(x_fake_noise)
    x_real_norm = log_norm(x_real)
    loss_norm = ((torch.nn.ReLU()(torch.abs(x_fake_norm - x_real_norm) - args.norm_bias))**2).mean()
    
    # F0 loss
    loss_f0 = f0_loss(F0_fake_noise, F0_real)
    
    # style F0 loss (style initialization)
    if x_refs is not None and args.lambda_f0_sty > 0 and not use_adv_cls:
        F0_sty, _, _ = nets.f0_model(x_ref)
        loss_f0_sty = F.l1_loss(compute_mean_f0(F0_fake_noise), compute_mean_f0(F0_sty))
    else:
        loss_f0_sty = torch.zeros(1).mean()
    
    # ASR loss
    loss_asr = F.smooth_l1_loss(ASR_fake_noise, ASR_real)
    
    # style reconstruction loss
    s_pred = nets.style_encoder(x_fake_noise, y_trg)
    loss_sty = torch.mean(torch.abs(s_pred - s_trg))
    
    # diversity sensitive loss
    if z_trgs is not None:
        s_trg2 = nets.mapping_network(z_trg2, y_trg)
    else:
        s_trg2 = nets.style_encoder(x_ref2, y_trg)
    x_fake2 = nets.generator(x_noise, s_trg2, masks=None, F0=GAN_F0_noise)
    x_fake2 = x_fake2.detach()
    _, GAN_F0_fake2, _ = nets.f0_model(x_fake2)
    loss_ds = torch.mean(torch.abs(x_fake_noise - x_fake2))
    loss_ds += F.smooth_l1_loss(GAN_F0_fake_noise, GAN_F0_fake2.detach())
    
    # cycle-consistency loss
    s_org = nets.style_encoder(x_noise, y_org)
    x_rec = nets.generator(x_fake_noise, s_org, masks=None, F0=GAN_F0_fake_noise)
    loss_cyc = torch.mean(torch.abs(x_rec - x_real))
    # F0 loss in cycle-consistency loss
    if args.lambda_f0 > 0:
        _, _, cyc_F0_rec = nets.f0_model(x_rec)
        loss_cyc += F.smooth_l1_loss(cyc_F0_rec, cyc_F0_real)
    if args.lambda_asr > 0:
        ASR_recon = nets.asr_model.get_feature(x_rec)
        loss_cyc += F.smooth_l1_loss(ASR_recon, ASR_real)
    
    # adversarial classifier loss
    if use_adv_cls:
        out_de = nets.discriminator.classifier(x_fake_noise)
        loss_adv_cls = F.cross_entropy(out_de[y_org != y_trg], y_trg[y_org != y_trg])
    else:
        loss_adv_cls = torch.zeros(1).mean()`

**Comment by yl4579** [comment]
@Charlottecuc Sorry for the late reply. I was pretty busy at the end of the year. You can make all `x_input` corrupted, but I'd recommend you set each transformation with a probability of 0.3, so there will be some samples that are not corrupted.

**Comment by Charlottecuc** [comment]
> The normalization may have possibly amplified the noises slightly, yet the point of log mel spec is actually the opposite: it tries to emphasize the speech instead of the noise. Probably the arbitrary mean and standard deviation may have some side effect, but if during training you make the model noise-robust, it should have no problems taking noisy input. As mentioned earlier [here](https://github.com/yl4579/StarGANv2-VC/issues/6), if you corrupt your input with [Audiomentations](https://github.com/iver56/audiomentations), it should have no problems dealing with noisy input. Just make sure you separate your `x_real` and `x_input` and only make `x_input` noisy.

@yl4579 Thank you for your reply. Just to make sure, when you mean "making the model noise-robust during training", do you mean only corrputing the input sof the cycle-consistency loss of generator, or, corrupting the inputs of the whole adversial training process (e.g. adding something like "denoising loss" to make the discriminator capable of classifing between clean and noisy inputs and force the generator to produce clean outputs)? Could you give more details?

Thank you very much.

**Comment by StyleTTS** [comment]
@Charlottecuc I'm sorry for the late reply because this issue was closed and I didn't get any notification. Not sure if it has been resolved, but what I meant was simply corrupting the input to the encoder but asking the model to reconstruct the clean (uncorrupted) version.

**Comment by yl4579** [comment]
@Charlottecuc I'm sorry for the late reply because this issue was closed and I didn't get any notification. Not sure if it has been resolved, but what I meant was simply corrupting the input to the encoder but asking the model to reconstruct the clean (uncorrupted) version.

**Comment by MMMMichaelzhang** [comment]
mel_tensor = (torch.log(1e-5 + mel_tensor) - self.mean) / self.std
have you fixed the noise problem? when i change mean =0, std =1 the noise gone,but it is too loud. @Charlottecuc

**Comment by skol101** [comment]
> The normalization may have possibly amplified the noises slightly, yet the point of log mel spec is actually the opposite: it tries to emphasize the speech instead of the noise. Probably the arbitrary mean and standard deviation may have some side effect, but if during training you make the model noise-robust, it should have no problems taking noisy input. As mentioned earlier [here](https://github.com/yl4579/StarGANv2-VC/issues/6), if you corrupt your input with [Audiomentations](https://github.com/iver56/audiomentations), it should have no problems dealing with noisy input. Just make sure you separate your `x_real` and `x_input` and only make `x_input` noisy.
@yl4579 
Please, help me here:
where's x_input actually? There's only x_real in trainer.py
```
x_real, y_org, x_ref, x_ref2, y_trg, z_trg, z_trg2 = batch
```

**Comment by yl4579** [comment]
@skol101 You need to pass in a noisy version here, call it `x_input`. The `x_input` is processed in `meldataset.py` with noises and reveberations.

**Comment by skol101** [comment]
I see, because I thought reverb and noise should be added right in StyleEncoder as per https://github.com/yl4579/StarGANv2-VC/issues/6#issuecomment-1103393827

**Comment by Kristopher-Chen** [comment]
> @Charlottecuc I'm sorry for the late reply because this issue was closed and I didn't get any notification. Not sure if it has been resolved, but what I meant was simply corrupting the input to the encoder but asking the model to reconstruct the clean (uncorrupted) version.

@Charlottecuc @yl4579 Are the noisy inputs added only when training the generator, or both the generator and the discriminator? Thank you!

**Comment by skol101** [comment]
> @Charlottecuc I'm sorry for the late reply because this issue was closed and I didn't get any notification. Not sure if it has been resolved, but what I meant was simply corrupting the input to the encoder but asking the model to reconstruct the clean (uncorrupted) version.

Style encoder is being called several times in the generator, but the only time it's called with x_real param is in the cycle consistency loss. So I guess that's there x_input should be used (but only 30% of the time). 

What do you think @Charlottecuc ?

**Comment by skol101** [comment]
I'm either doing something wrong or adding reverbs and background noises does nothing
```python
# cycle-consistency loss
    s_org = nets.style_encoder(x_input, y_org)
    x_rec = nets.generator(x_fake, s_org, masks=None, F0=GAN_F0_fake)
    loss_cyc = torch.mean(torch.abs(x_rec - x_real))
```
In meldataset.py
```python
 def _load_tensor(self, data, corrupt_x_input=False):
        wave_path, label = data
        label = int(label)
        wave, sr = sf.read(wave_path)

        if random.uniform(0, 1) <= 0.3 and corrupt_x_input:
            augmenter = Compose(
                [
                    RoomSimulator(
                        p=0.8,
                        leave_length_unchanged=True,
                    ),
                    AddBackgroundNoise(
                        sounds_path=BACKGROUND_NOISE_FILES,
                        min_snr_in_db=20,
                        max_snr_in_db=35,
                        p=0.5,
                    )
                ]
            )
            try:
                wave = augmenter(samples=wave, sample_rate=sr)
            except IndexError as error:
                print('error index error with wav file', wave_path)
            except ValueError as errorValue:
                print('error value error with wav file', wave_path)
        wave_tensor = torch.from_numpy(wave).float()
        return wave_tensor, label
```

**Comment by skol101** [comment]
Wow, is this really a mystery @yl4579  ?

**Comment by Charlottecuc** [comment]
_> Here it was reported that added reverb/background noises did help [#6 (comment)](https://github.com/yl4579/StarGANv2-VC/issues/6#issuecomment-1108044051)
> 
> Maybe the solution is to denoise the input wave before proceeding with inference, so something like Facebook denoiser can be used, but this suggestion points to using noise trained vocoder [#6 (comment)](https://github.com/yl4579/StarGANv2-VC/issues/6#issuecomment-1099653031)

If you would like to train an any-to-any model, then add data aug to style encoder will help. 
If you denoise the source wave before inference, some distortion caused by noise will disappear, but new issue will come, since most denosing models will weaken the voice when eliminating noises. The problem was confirmed in many VC papers.
Training denoising HiFi-GAN might help, but may not achieve what you expect for VC issues. Because when inferencing with noisy wave, there might be some mispronunciations in converted mels, which can not be post-corrected by vocoder.

**Comment by yl4579** [comment]
@Charlottecuc Sorry I'm pretty busy with my other paper submissions so I can't join the discussion at this point, but I have reopened the issue for further discussion and will provide some feedback after I finished my work.

**Comment by yl4579** [comment]
@Charlottecuc I do have some time now to discuss this problem. I have noticed similar problems with noisy input and have not yet come up with a good solution. The major problem with the GAN-based model is that it is difficult to design denoise loss functions because the target is not as clear as in PPG or TTS based VC models (in that case you have L1 reconstruction loss directly). Not sure if you have got any good solution to this problem, but I would suggest adding some noises in the time-frequency domain by reverse mel-scale and recomputing the mel scale (or you can train a model end-to-end if you prefer). 

The key here is to add noise to the converted speech and force the model to convert the converted speech back to the clean output. Because one problem I noticed is that even if you add noise to the input during training, the model does not produce good converted examples sometimes. It somehow finds a way to trick the loss function so that the converted speech is clear, but the second time conversion to the source domain works quite well so the cycle consistency loss is still low. Adding noises to the converted results force the model to denoise the noisy speech directly. Another way is to add a denoise loss directly where the input is a noisy speech and the source style vector and the output is a clean speech. This might make the model overfit however so the converted speech might not sound similar to the target. This is in general a very difficult topic so there's still a lot of work to be done.

**Comment by skol101** [comment]
This does a pretty good job of removing noises from the speech https://github.com/Rikorose/DeepFilterNet

**Comment by mayank-git-hub** [comment]
Another approach which works is to first train the model on a clean dataset and once the model is trained, freeze the model parameters and add two enhancement blocks to the encoder and the style encoder to enhance the noisy voice
in the feature domains using synthetically distorted data. We use the embeddings extracted from clean samples by
the original frozen encoder as targets and train the newly added enhancement blocks by minimizing the L1 distance between the target and the outputs obtained by the encoders with enhancement blocks from the distorted samples.

You can refer to our paper https://arxiv.org/pdf/2210.11096.pdf which shows the figures and results on distorted/noisy samples using StarGANv2-vc model architecture.

**Comment by Patelraj8694** [comment]
Hi @mayank-git-hub , I have similar application to your idea and I want speech conversion whisper or distorted speech. I do not have that much knowledge in fine tuning model , can you help me out ?
